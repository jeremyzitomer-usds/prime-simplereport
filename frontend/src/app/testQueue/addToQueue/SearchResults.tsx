import React, { useState } from "react";
import Button from "../../commonComponents/Button";
import AoeModalForm from "../AoEForm/AoEModalForm";
import { displayFullName } from "../../utils";

interface SearchResultsProps {
  patients: any[]; //TODO TYPE: Patient
  onAddToQueue: (a: any, b: any) => null; //TODO TYPE: Patient, answers
  facilityId: string;
  patientsInQueue: string[];
}

const SearchResults = ({
  patients,
  onAddToQueue,
  facilityId,
  patientsInQueue,
}: SearchResultsProps) => {
  const [dialogPatient, setDialogPatient] = useState(null);

  if (patients.length === 0) {
    return <h3> No results </h3>;
  }
  return (
    <div className="usa-card__container shadow-3 results-dropdown">
      <div className="usa-card__body">
        {dialogPatient !== null && (
          <AoeModalForm
            saveButtonText="Continue"
            patient={dialogPatient}
            onClose={() => setDialogPatient(null)}
            saveCallback={(a: any) => onAddToQueue(dialogPatient, a)}
            facilityId={facilityId}
          />
        )}
        <table className="usa-table usa-table--borderless">
          <thead>
            <tr>
              <th scope="col">Full name</th>
              <th scope="col">Date of Birth</th>
              <th scope="col">Unique ID</th>
              <th scope="col">Actions</th>
            </tr>
          </thead>
          <tbody>
            {patients.map((p) => (
              <tr key={p.internalId}>
                <td>
                  {displayFullName(p.firstName, p.middleName, p.lastName)}
                </td>
                <td>{p.birthDate}</td>
                <td>{p.lookupId}</td>
                <td>
                  {patientsInQueue.indexOf(p.internalId) === -1 ? (
                    <Button
                      variant="unstyled"
                      label="Begin Test"
                      onClick={() => setDialogPatient(p)}
                    />
                  ) : (
                    "Test in progress"
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default SearchResults;
