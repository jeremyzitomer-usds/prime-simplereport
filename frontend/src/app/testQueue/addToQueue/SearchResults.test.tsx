import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import SearchResults from "./SearchResults";

//TODO: TYPES
const patients /* : Partial<Patients> */ = [
  {
    internalId: "a123",
    firstName: "George",
    middleName: "",
    lastName: "Washington",
  },
  {
    internalId: "b456",
    firstName: "Martin",
    middleName: "Luther",
    lastName: "King",
  },
  {
    internalId: "c789",
    firstName: "Barack",
    middleName: "Hussein",
    lastName: "Obama",
  },
];

describe("SearchResults", () => {
  it("should say 'No Results' for no matches", () => {
    const component = renderer.create(
      <SearchResults
        patients={[]}
        patientsInQueue={[]}
        facilityId={"1234-5678"}
        onAddToQueue={jest.fn()}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should show matching results", () => {
    const component = renderer.create(
      <SearchResults
        patients={patients}
        patientsInQueue={[]}
        facilityId={"1234-5678"}
        onAddToQueue={jest.fn()}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("links the non-duplicate patient", () => {
    const addToQueue = jest.fn();
    render(
      <SearchResults
        patients={patients}
        patientsInQueue={["a123", "c789"]}
        facilityId={"1234-5678"}
        onAddToQueue={addToQueue}
      />
    );

    expect(screen.getAllByText("Test in progress")).toHaveLength(2);
    expect(screen.getAllByText("Begin Test")).toHaveLength(1);
  });
});
