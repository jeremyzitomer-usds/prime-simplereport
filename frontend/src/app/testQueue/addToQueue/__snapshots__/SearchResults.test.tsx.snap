// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should say 'No Results' for no matches 1`] = `
<h3>
   No results 
</h3>
`;

exports[`SearchResults should show matching results 1`] = `
<div
  className="usa-card__container shadow-3 results-dropdown"
>
  <div
    className="usa-card__body"
  >
    <table
      className="usa-table usa-table--borderless"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Full name
          </th>
          <th
            scope="col"
          >
            Date of Birth
          </th>
          <th
            scope="col"
          >
            Unique ID
          </th>
          <th
            scope="col"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            George Washington
          </td>
          <td />
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin Test
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Martin Luther King
          </td>
          <td />
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin Test
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Barack Hussein Obama
          </td>
          <td />
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin Test
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
